src/rust/collections/concurrent_ring.rs

```diff
@@ -0,0 +1,765 @@
+        let pop_offset: usize = peek(self.pop_offset);
+        //println!("{:?} {:?}", push_offset, pop_offset);
+        self.available_space(push_offset, pop_offset)
+    }
+
+    /// Attempts to insert a buffer of [len] bytes into the ring buffer.
+    pub fn try_push(&self, buf: &[u8]) -> Result<usize, Fail> {
+        let len: usize = buf.len();
+        if len == 0 {
+            return Err(Fail::new(libc::EINVAL, "Buffer must be non-zero length"));
+        }
+        // reserve_space will allocate space for the header.
+        if let Some(push_offset) = self.reserve_space(len) {
+            debug_assert!(push_offset % 2 == 0);
+            // Push first part of buffer. If longer than the capacity of the ring, wrap around.
+            let first_offset: usize = push_offset + HEADER_SIZE;
+            let first_len: usize = if push_offset + len + HEADER_SIZE > self.capacity() {
+                self.capacity() - first_offset
+            } else {
+                len
+            };
+            let buf_ptr: *const u8 = buf.as_ptr();
+            let ring_ptr: *mut u8 = unsafe { self.buffer.get_mut().as_mut_ptr() };
+            // Copy the data into the ring buffer.
+            unsafe {
+                copy(buf_ptr, ring_ptr.add(first_offset), first_len);
+            }
+            // If there is remaining data in the buffer, wrap around.
+            if len > first_len {
+                // Copy the data into the ring buffer.
+                unsafe {
+                    copy(buf_ptr.add(first_len), ring_ptr, len - first_len);
+                }
+            }
+            // Commit the write by atomically writing the header to release the buffer. The overwritten header MUST be
+            // 0. The header describes just the length of the payload.
+            let old: usize = self.write_header(push_offset, len);
+            debug_assert_eq!(old, 0);
+            Ok(len)
+        } else {
+            Err(Fail::new(libc::EAGAIN, "No space in the ring buffer"))
+        }
+    }
+
+    /// Inserts an item at the enqueue of the target ring buffer. This function may block (spin).
+    #[allow(unused)]
+    pub fn push(&self, buf: &[u8]) -> Option<usize> {
+        loop {
+            match self.try_push(buf) {
+                Ok(len) => return Some(len),
+                Err(Fail { errno, cause: _ }) if retry_errno(errno) => continue,
+                Err(e) => return None,
+            }
+        }
+    }
+
+    /// Attempts to remove next message from the ring buffer up to [len] bytes and copies into [buf]. This function
+    /// does not block.
+    pub fn try_pop(&self, buf: &mut [u8]) -> Result<usize, Fail> {
+        let len: usize = buf.len();
+        if len == 0 {
+            return Err(Fail::new(libc::EINVAL, "Buffer must be non-zero length"));
+        }
+
+        // Is this safe or do we need an atomic indirect read?
+        let pop_offset: usize = peek(self.pop_offset);
```

Comment:
Having thought about this more, I am fairly sure that this might be an issue with concurrent writers. Let me think of a fix

Reply:
Let me know what you suggest for this.